### DISCUSSION TITLE
Add functions in the level trajectory

### DISCUSSION THREAD

**Discussion by Felienne** [hedyorg/hedy#4036] [Archive]
We are getting more and more requests to also add functions to the levels and I am inclined to take those requests seriously. A few reasons:

* More and more older students are using Hedy, even university level students. That makes total sense because programming is as hard for older learners as it is for younger ones, but that means that they will probably need a bit of power. 
* Because of the localization, saying "just learn this is Python" is a lot less viable now for non-English users

Question is of course.... where in the trajectory? I am inclined to say let's do it before `while` loops? While `while` adds some power it is not a lot of power compared to `for` and it also hard. But putting a level in between is quite involved because we have the shift content (but we have "chopped with that axe before" so it should be doable)

**Comment by jpelay** [comment]
I really like this idea! I think this can be a worthwhile. I often think about this since there is a point in Hedy (around level 14) where the adventures become more and more complex, where the addition of functions can make things clearer, for example in the Tic Tac Toe story.  I think functions are not supper difficult as to merit their own level, but being in the same level as `while` and `for` might be a tall ask.

Also, I understand that this is a little tangential to this discussion so maybe we should open another thread for this, but you mentioned that older students are starting to use Hedy, but the Hedy content is very much tailored towards K-12. Do you think is worth adding more content for those older students?

**Reply by Felienne** [reply]
> I really like this idea! I think this can be a worthwhile. I often think about this since there is a point in Hedy (around level 14) where the adventures become more and more complex, where the addition of functions can make things clearer, for example in the Tic Tac Toe story. I think functions are not supper difficult as to merit their own level, but being in the same level as `while` and `for` might be a tall ask.

Yes I think so too! Let's discuss it is a meeting, there is no real rush!
 
> Also, I understand that this is a little tangential to this discussion so maybe we should open another thread for this, but you mentioned that older students are starting to use Hedy, but the Hedy content is very much tailored towards K-12. Do you think is worth adding more content for those older students?

Yeah, I will make a new discussion so we have a place to discuss that!

**Comment by jlole** [comment]
I will be working on this for my bachelor's thesis. Before reading this discussion, I was also leaning towards implementing functions around Level 15 (while loops).

We should also decide how we will make the concept of functions gradual and then decide where to add each of those levels. My idea is:

### 1. Simple function declarations and calls
**example code**
```
define my_function
	print "inside function"
call my_function
```	
output:
`inside function`

### 2. Adding arguments
**example code**
```
define sum using a,b
	print a + b
ten is 10
call sum with 5, ten
```	
output:
`15`

### 3. Return value and parentheses
**example code**
```
define sum(a,b)
	return a + b
ten is 10
print sum(5, ten)
```	
output:
`15`

### 4. Semicolon and def
**example code**
```
def sum(a,b):
	return a + b
ten is 10
print sum(5, ten)
```	
output:
`15`

One thing I'm not sure about is the parentheses. Currently they are added in the final level, but with functions, it might be very tricky to not use them when calling the function with arguments to be used within another command e.g.
parsing `print sum(4,5)`
vs parsing `print sum with 4,5`

Adding the colon is currently introduced in level 17 so we can add it for functions in the same level.

I would like to hear your thoughts on this @Felienne, also if the keywords should be different, not sure which are the best for students to understand easily. Eventually the only remaining keyword will be `def`

**Reply by Felienne** [reply]
> I will be working on this for my bachelor's thesis. Before reading this discussion, I was also leaning towards implementing functions around Level 15 (while loops).

Yes, I would say before even.
> 
> We should also decide how we will make the concept of functions gradual and then decide where to add each of those levels. My idea is:
> 
> ### 1. Simple function declarations and calls
> **example code**
> 
> ```
> define my_function
> 	print "inside function"
> call my_function
> ```
> 
> output: `inside function`
> 
> ### 2. Adding arguments
> **example code**
> 
> ```
> define sum using a,b
> 	print a + b
> ten is 10
> call sum with 5, ten
> ```
> output: `15`

I love this idea and this syntax too, using and with are very elegant and sufficiently different from each other (rather than brackets for both as non-learning languages)

> One thing I'm not sure about is the parentheses. Currently they are added in the final level, but with functions, it might be very tricky to not use them when calling the function with arguments to be used within another command e.g. parsing `print sum(4,5)` vs parsing `print sum with 4,5`

Yes I see but we have made many compromises in "not so easy parsing" in favor of a more usable language, so I am leaning towards keeping the brackets at 18. We could for example not allow nested function calls but only the assigning of a function result to a variable. That is probably more learnable too!

> Adding the colon is currently introduced in level 17 so we can add it for functions in the same level.

Yes that is easy enough!
 
> I would like to hear your thoughts on this @Felienne, also if the keywords should be different, not sure which are the best for students to understand easily. Eventually the only remaining keyword will be `def`

Yes I guess we'd do that in 18 too.

Very excited to see this idea move on!!
